src/styles/Styled.js

```diff
@@ -0,0 +1,330 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import warning from 'warning';
+import getDisplayName from 'recompose/getDisplayName';
+import contextTypes from 'react-jss/lib/contextTypes';
+import { create } from 'jss';
+import jssGlobal from 'jss-global';
+import jssNested from 'jss-nested';
+import jssCamelCase from 'jss-camel-case';
+import jssDefaultUnit from 'jss-default-unit';
+import jssVendorPrefixer from 'jss-vendor-prefixer';
+import jssPropsSort from 'jss-props-sort';
+import * as ns from 'react-jss/lib/ns';
+import createMuiTheme from './createMuiTheme';
+import themeListener from './themeListener';
+import createGenerateClassName from './createGenerateClassName';
+import getStylesCreator from './getStylesCreator';
+
+export const preset = () => ({
+  plugins: [
+    jssGlobal(),
+    jssNested(),
+    jssCamelCase(),
+    jssDefaultUnit(),
+    jssVendorPrefixer(),
+    jssPropsSort(),
+  ],
+});
+
+// New JSS instance.
+const jss = create(preset());
+
+// Use a singleton or the provided one by the context.
+const generateClassName = createGenerateClassName();
+
+// Global index counter to preserve source order.
+// As we create the style sheet during componentWillMount lifecycle,
+// children are handled after the parents, so the order of style elements would
+// be parent->child. It is a problem though when a parent passes a className
+// which needs to override any childs styles. StyleSheet of the child has a higher
+// specificity, because of the source order.
+// So our solution is to render sheets them in the reverse order child->sheet, so
+// that parent has a higher specificity.
+let indexCounter = Number.MIN_SAFE_INTEGER;
+
+export const sheetsManager: Map<*, *> = new Map();
+
+// We use the same empty object to ref count the styles that don't need a theme object.
+const noopTheme = {};
+
+// In order to have self-supporting components, we rely on default theme when not provided.
+let defaultTheme;
+
+function getDefaultTheme() {
+  if (defaultTheme) {
+    return defaultTheme;
+  }
+
+  defaultTheme = createMuiTheme();
+  return defaultTheme;
+}
+
+class Styled extends React.Component {
+  static defaultProps = {
```

Comment:
I have been removing all the static properties as they get in the way of refactoring components from classes to functions.

Reply:
Don't do that if it is still a class, it is invalid typescript and could harm our potential conversion. If it's a class, static props are the right way, the other way is just a hack after all.

Reply:
Oh, TypeScript doesn't support this?

Reply:
You have to: augment the definition of the target, or cast it. So it supports it, but you have to cast it to the type first. In other words it is uglier on the outside, vs static class prop.

Reply:
Alright. At least I can keep using this pattern at work. It's saving us time refactoring component `class <-> function`.